net: tls: Add DTLS support to secure sockets

rlubos opened the pull request:
This PR adds support to DTLS with sockets API.

A new socket option was added to select DTLS role. A user should specify on a socket whether it should behave as a DTLS server or client (by default it's a client). Of course all options implemented for TLS can be used with DTLS as well.

For starters, DTLS server supports only one connection at a time. If needed, it could be extened in the future to support multiple client connections, yet it would require more sophisticated resource and datagram management (each connection would need it's own `mbedtls_ssl_context`).

The handshake procedure is almost transparent to the user. It will take place during initial call to `sendto`/`recvfrom`. DTLS client will report an error on handhshake failure. DTLS server will silently ignore the error and prepare TLS context for another handshake. 

Because DTLS requires extra resources per TLS context and some extra code, it can be disabled by Kconfig if not used.

As DTLS proven itself to be trickier than TLS, there are quite a lot of changes. I've tried to keep them split in a logically separated commits, to keep reviewing simple. Still, let me know if it needs any refactoring. 

As for the samples, DTLS support was added for socket `echo_client`/`echo_server`. By default, the samples will run both TLS and DTLS simultaneously, 4 connections in total. They can be tested with `dtls-client`/`dtls-server` from `net-tools` as well.

rlubos committed:
net: tls: Add DTLS protocol types

Define DTLS protocol types and and Kconfig option to enable DTLS
support.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Add socket option to set DTLS role

Add write-only socket option to set role for DTLS connection. This
option is irrelevant for TLS connections.

This options accepts and integer with a TLS role, compatible with
mbedTLS values:
0 - client,
1 - server.

By default, DTLS will assume client role.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Add timer handling for DTLS

Add timer functions required by DTLS.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Add DTLS peer address helpers

Add helper functions to handle stored DTLS peer address.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Add cookie for DTLS

Add DTLS cookie to TLS context, required by DTLS servers.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: socket: Extract flag related functions to internal header file

Functions for checking flags set on sockets are needed by TLS sockets as
well, therefore extract them to a separate header file to avoid code
duplication.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Add DTLS bio functions

Add binary IO functions for DTLS connections.

dtls_rx function is more complex than it's TLS counterpart due to fact,
that DTLS does not allow blocking operation for this function. A simple
timeout mechanism was implmented basing on the zsock_poll function.

This function also verifies peer address. As currently only a single
DTLS connection is supported on a socket, if a DTLS connection is
established, and we receive datagram from different peer, it is silently
dropped.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Handle HELLO_VERIFY_REQUIRED during DTLS handshake

DTLS handshake can return MBEDTLS_ERR_SSL_HELLO_VERIFY_REQUIRED, which
indicate that TLS session context should be reset.

Also, store information whether TLS connection has beed established.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Adapt connect for DTLS

Extend connect function with DTLS functionality.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Support DTLS send and handshake

Add support for DTLS send/sendto function.

For DTLS clients, send function will try to establish DTLS connection
before sending data. If DTLS handshake fails, it will return an error.

For DTLS servers, send function requires to have DTLS connection already
established.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Support DTLS recv and handshake

Add support for DTLS recv/recvfrom function.

For DTLS client, recv function requires to have an already established
DTLS connection.

For DTLS servers, this function will try to establish DTLS connection
before receiving data. In case that DTLS handshake fails, recv function
will silently retry.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Add timeout for mbetTLS read

Specify timeout value for mbedtls_ssl_read function for DTLS servers.
Adding this can prevent TLS context lockup in case blocking recv is used
and peer has shut down DTLS connection without closing it gracefully.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Make TLS poll implementation check mbedTLS data

Make TLS poll function verify if decrypted data is available after
socket has notified activity with POLLIN flag. This prevents from giving
false notifications in case data was received on socket but was consumed
by mbedTLS.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
samples: net: Add DTLS support to socket echo_client/echo_server

Add DTLS support to socket echo_client and echo_server samples.

Additionally, move TLS-related configs to overlay-tls.conf config file,
to align with other examples.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tls: Enable non-blocking DTLS handshake

Introduce non-blocking DTLS handshake, used during recv function call.
This prevents from blocking while waiting for initial handshake packet
on non-blocking sockets during receive.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -38,6 +38,14 @@ config NET_SOCKETS_SOCKOPT_TLS
 
+config NET_SOCKETS_ENABLE_DTLS
+	bool "Enable DTLS socket support [EXPERIMENTAL]"
+	default n
```
No need to set the "default n" as the value is false automatically.

jukkar commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -350,6 +356,56 @@ static int tls_release(struct tls_context *tls)
+		return false;
+	}
+
+	if (peer_addr->sa_family == AF_INET6) {
```
We could optimize this a bit more, so something like this

`if (IS_ENABLED(CONFIG_NET_IPV6) && peer_addr->sa_family == AF_INET6) {`

jukkar commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -350,6 +356,56 @@ static int tls_release(struct tls_context *tls)
+
+		return (addr1->sin6_port == addr2->sin6_port) &&
+			net_ipv6_addr_cmp(&addr1->sin6_addr, &addr2->sin6_addr);
+	} else if (peer_addr->sa_family == AF_INET) {
```
Similar check also for IPv4

jukkar commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -404,16 +405,116 @@ static void dtls_peer_address_get(struct net_context *context,
+
+static int dtls_rx(void *ctx, unsigned char *buf, size_t len, uint32_t timeout)
+{
+	int sock = POINTER_TO_INT(ctx);
```
Could you rearrange the parameters in this function as we usually have the variables that are set a value in the beginning.

jukkar commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -404,16 +405,116 @@ static void dtls_peer_address_get(struct net_context *context,
+					return err;
+				}
+			} else {
+			/* For clients it's incorrect to receive when no peer
```
Indentation is wrong

jukkar commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -643,7 +743,17 @@ static int tls_mbedtls_init(struct net_context *context, bool is_server)
 
-	mbedtls_ssl_set_bio(&context->tls->ssl, context, tls_tx, tls_rx, NULL);
+	if (type == MBEDTLS_SSL_TRANSPORT_STREAM) {
+		mbedtls_ssl_set_bio(&context->tls->ssl, context,
```
Just pondering here that should we have a way to disable TLS but leave DTLS enabled. This could happen if TCP is not enabled in Kconfig. Would this be a big change in the code?

rlubos replied:
Well, we could add another switch to enable/disable TLS itself, yet the benefit would be very little. There are some functions in this module that are TLS specific (like `tls_rx`/`tlx_tx` bio functions), that could be compiled out, but that's it I suppose.

DTLS is based on TLS (I guess we could consider it an extension) and technically from mbedTLS side, most (if not all) TLS features are needed for DTLS operation.

As for the disabled TCP, in that case TLS socket creation will simply fail even at this point, no extra switches needed.

jukkar replied:
I was thinking a case where TCP is disabled, in which case TCP specific code in various TLS libraries could be compiled out. Anyway, we can ponder this more later, no need to start optimize this kind of thing here.

jukkar commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -64,6 +64,9 @@ struct tls_context {
 	/** Socket flags passed to a socket call. */
```
Typo in commit message of "net: tls: Support DTLS send and handshake".
s/failes/fails/

jukkar reviewed (changes requested):
Thanks Robert, overall looks good, minor tweaks needed.

jukkar commented:
@pfalcon do you have any comments to this PR?

jukkar commented:
This was meant for 1.13

ceolin commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -733,6 +736,30 @@ static int tls_opt_peer_verify_set(struct net_context *context,
+	int *role;
+
+	if (!optval) {
+		return -EFAULT;
```
As this is a parameter, shouldn't be better using -EINVAL ?

pfalcon replied:
+1. EFAULT means "would lead to MMU fault when accessed", and is usually returned for arbitrary stray pointers. POSIX systems usually don't return EFAULT for NULL pointers, use EINVAL instead.

rlubos replied:
Will fix.

ceolin commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -733,6 +736,30 @@ static int tls_opt_peer_verify_set(struct net_context *context,
+
+	if (optlen != sizeof(int)) {
+		return -EINVAL;
+	}
```
why this check ? optlen is a socklen_t that is a typedef to size_t. If size_t happens to be different of int (or unsigned int) this will fail. If you really need something of int's size, shouldn't be better change this parameter's type ?

ceolin commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -156,6 +168,44 @@ static int tls_entropy_func(void *ctx, unsigned char *buf, size_t len)
+	elapsed_ms = k_uptime_get_32() - timing->snapshot;
+
+	if (elapsed_ms >= timing->fin_ms) {
+		return 2;
```
Using constants for these return values would make it better to read

jukkar replied:
These values come directly from mbedtls https://tls.mbed.org/api/ssl_8h.html#adea005600b532d788c05e9d01b42895b

I also used literal values in net_app in same function. There could be a comment here describing the values:
```
This callback must return:
       -1 if cancelled (fin_ms == 0),
        0 if none of the delays have passed,
        1 if only the intermediate delay has passed,
        2 if the final delay has passed.
```

ceolin replied:
As this is an internal thing it's ok :)  Add this comment explaining these values is really nice !

rlubos replied:
I'll add comment claryfing this.

ceolin commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -350,6 +356,57 @@ static int tls_release(struct tls_context *tls)
+{
+	socklen_t len = min(context->tls->dtls_peer_addrlen, *addrlen);
+
+	memcpy(peer_addr, &context->tls->dtls_peer_addr, len);
```
couldn't just do:
```
memcpy(peer_addr, &context->tls->dtls_peer_addr, sizeof(struct sockaddr));
*addrlen = context->tls->dtls_peer_addrlen
```

?

rlubos replied:
Well, not exactly, because size of the buffer provided could be smaller than `sizeof(struct sockaddr)`.

This function was intended to use with `recvfrom` and according to it's manpage:
```
If  the  actual length of the address is greater than the length of the supplied sockaddr structure, the stored address shall be truncated.
```

That's why it's implemented like this and I think we should keep it this way.

ceolin commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1143,6 +1145,9 @@ int ztls_connect(int sock, const struct sockaddr *addr, socklen_t addrlen)
 			}
 
+			/* Do not use any socket flags during the handshake. */
+			context->tls->flags = 0;
```
is it ok override flags here and restore after handshake ?

rlubos replied:
The flags are set explicitly during each `send`/`recv' call. As for `connect`/`accept` there is no way to pass flags to a socket, but we still use it during the handshake, therefore we should clear any leftover flags. Still, at this point it's more like an extra precaution, as TLS socket is currently not reusable and should be closed after connection is closed.

ceolin commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1197,6 +1202,9 @@ int ztls_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
 		}
 
+		/* Do not use any socket flags during the handshake. */
+		child_context->tls->flags = 0;
```
ditto

rlubos committed:
net: tls: Use EINVAL for NULL pointer error in option setters

Originally EFAULT was used to indicate NULL pointer error in TLS option
set/get functions. EINVAL was suggested to be more apropriate error code
for this case, hence replace it.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@ceolin Thank you for the feedback, I've pushed the updated branch.

pfalcon commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1236,31 +1244,126 @@ ssize_t ztls_recv(int sock, void *buf, size_t max_len, int flags)
 	return ztls_recvfrom(sock, buf, max_len, flags, NULL, 0);
 }
 
+static ssize_t sendto_tls(struct net_context *context, const void *buf,
```
Why is this called "sendto", if it doesn't accept dest addr param?

rlubos replied:
My intention was to keep the naming uniform, as this static function is part of the `sendto` function, but sure I can rename it. Same with others.

pfalcon replied:
I see, but "sendto" and "recvfrom" socket verbs have a very specific meaning of including an extra arg param, and it may get really confusing if they don't, so I appreciate fixing that.

pfalcon commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1236,31 +1244,126 @@ ssize_t ztls_recv(int sock, void *buf, size_t max_len, int flags)
+		return -1;
+	}
+
+	/* Check if not trying to send to other peer, which we're
```
This comment is hard to read, because it has double negation ;-).

rlubos replied:
Agreed, it might not be straightforward, will repharse.

pfalcon commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1366,27 +1366,11 @@ ssize_t ztls_sendto(int sock, const void *buf, size_t len, int flags,
 
-ssize_t ztls_recvfrom(int sock, void *buf, size_t max_len, int flags,
-		      struct sockaddr *src_addr, socklen_t *addrlen)
+static ssize_t recvfrom_tls(struct net_context *context, void *buf,
```
Same comment as for sendto_tls - why recvfrom, if no src addr storage ptr?

pfalcon commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1406,12 +1390,175 @@ ssize_t ztls_recvfrom(int sock, void *buf, size_t max_len, int flags,
+
+	switch (ret) {
+	case MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY:
+		ret = -ECONNRESET;
```
What this error code is based on?

rlubos replied:
Well, server closed connection so the socket is not usable anymore. From `recvfrom` man page we have:
```
ECONNRESET
              A connection was forcibly closed by a peer.
```
So I've figure that's the error code to use here.

pfalcon replied:
Strange, my recvfrom man page doesn't explicitly mention ECONNRESET (Ubuntu 16.04).

Let's however go thru what's known about ECONNRESET. First of all, "RESET" there unlikely a random word, but kinda tied to TCP's RST packet. A well-known way to get ECONNRESET is when peer closes socket connection on its side without having read all the data already sent by the originator. For example, a client makes an HTTP request, server send a complete page in return, but client reads it partially and closes the connection (e.g. quits in POSIX-like environment). Server will get ECONNRESET in this case. So, it's clear that ECONNRESET is tied with TCP processing. 

Thus, there would be 2 questions:

1. Is ECONNRESET possible with UDP sockets at all?
2. Even if it is, why do you think that "forcibly closed" applies here. Above is given example what is "forcible" or "abrupt" closure is - closing connection without reading all data peer already sent. In case of TLS, a forcible closure is closing the connection on socket level, instead of TLS level. The latter involves sending a closure notification, and that's exactly what you get here.

So, even ignoring first question, what we have here is clean, coordinated closure, not forced at all, so it can't be an error, and should be an EOF return (i.e. len=0) instead.

rlubos replied:
Ok, thanks for elaborating, I'll return 0 then.

pfalcon commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -1406,12 +1390,175 @@ ssize_t ztls_recvfrom(int sock, void *buf, size_t max_len, int flags,
+
+	case MBEDTLS_ERR_SSL_TIMEOUT:
+		(void)mbedtls_ssl_close_notify(&context->tls->ssl);
+		ret = -ECONNABORTED;
```
A timeout error wouldn't be a better choice?

rlubos replied:
Perhaps, will update.

pfalcon commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -43,6 +43,19 @@ config NET_SOCKETS_ENABLE_DTLS
 	help
 	  Enable DTLS socket support. By default only TLS over TCP is supported.
 
+config NET_SOCKETS_DTLS_TIMEOUT
```
TLS (and TCP) has the same problem in general. We need to offer more general solutions, e.g. implement SO_RCVTIMEO/SO_SNDTIMEO sockopts, while still allow to set specific default (instead of "no timeout") via Kconfig.

rlubos replied:
Does TLS have the same problem? What I wanted to protect against with this timeout is TLS resources lockup in case peer went down. TLS is protected by TCP in this case - the connection will time-out at TCP level and it'll be closed. UDP is missing such a feature, hence extra effort is needed to prevent TLS context lockup.

As per `mbedTLS`, read timeout is DTLS specific.

pfalcon replied:
> TLS is protected by TCP in this case - the connection will time-out at TCP level and it'll be closed.

Classical TCP doesn't have any timeouts, it was designed with hardcore reliability in mind. You could detect some "timeout" on send, but there's even not possible to detect timeout on recv in classical TCP. An example I like is that there can be recv() call waiting for a year. Then someone cuts the wires. They are broken for next 10 years. Then someone fixes them. recv() will be able to receive data from the other connected peer as if nothing happened.

There're *options* and *extensions* to TCP to address that. They're all particular, and not even (fully) implemented in Zephyr. And in general, only a specific application may know they best way to deal with timeouts.

> As per mbedTLS, read timeout is DTLS specific.

So, what we're doing here is taking specific protocols (TCP, UDP, TLS, DTLS), their specific implementations (Zephyr native IP stack, mbedTLS, etc.), and trying to expose them via the standardized POSIX Sockets API. My call is exactly about that: we should strive for all the following: a) provide adequate timeout support (as needed by most real-world apps); b) make it full coverage (not just DTLS for example); c) adjustable via standard Sockets API calls; d) still standards compliant (allowing for infinite timeout).

Nothing to do particularly here now of course, just a call for thinking and forthcoming changes along the lines above.

pfalcon reviewed (commented):
Due to @jukkar's hastiness, here comes *post*review, still raising some points worth attention IMHO.

pfalcon commented:
> For starters, DTLS server supports only one connection at a time. If needed, it could be extened in the future to support multiple client connections, yet it would require more sophisticated resource and datagram management (each connection would need it's own mbedtls_ssl_context).

So, layering violation catches and bites us again, eh? ;-)

Note that I can't claim to be a UDP usage expert, let alone DTLS one, so I learn as I go, plus some previous thinking (which isn't backed much by practical experience, so is "original research" in Wikipedia terms).

So, a feature of UDP is that it is both datagram **and** connectionless transport. Indeed, most well-known usages of UDP are "single request - single response" protocols, like DNS or NTP, which fit perfectly "connectionless". But sooner or later need for a conversation-like protocol over UDP may emerge, which means storing state, which effectively means going connection-oriented. And here UDP leaves that to application-level, there may be different choices, and again, I'm not sure about the "best practices" or at least well-known choices here. I can imagine at least 2 options:

1. Emulate TCP with its accept(). This means keeping "listening" UDP socket, but on first received pkt there create a new pkt and send reply from it. Expecting client to follow this switch, i.e. send initial request to a well-known port, but be ready to receive reply from another port, and then use that port for follow-up communication. This approach pushes client multiplexing down/back to the IP stack, which may be more performant, but may also use more resources. This choice would be seem to be good match for thread-based sever model: threads are already bloat, so having one socket for each thread would be also expected bloat, while would allow to avoid any resource sharing problems.
2. Multiplex by client address, fully on application-level. This duplicates multiplexing work done by the IP stack (which already multiplexes by port number), but might allow to save resources on creating socket objects just for purpose of multiplexing.

Now getting back to DTLS, how it is different from UDP is that it is inherently connection-oriented, as there's cipher state to keep for each connection (RFC6347 uses term "association"). Still, the RFC allows to keep as much UDP freedom as possible with the statement "It is designed to run in application space, without requiring any kernel modifications." For example, when implementing (or more exactly, integrating) DTLS on application level, it would be still possible to implement "accept-like" model (p.1) above. But pushing it under sockets API, only choice 2 is available.

But then yes, DTLS server implementation must track *all* clients, an implementation with just 1 connection is nothing but a toy (but a typical starting choice for Zephyr implementations, yeah). This means applying some (user-defined) constraints on the number of clients. YMMV, but (ab)using listen() call for this may be only better than defining bunch of sockopt's. Another problem is garbage collection of clients. All these problems would be easier if done on application level, but instead now need to find workarounds/define interfaces for all this, likely rather rigid (and at the same time probably already hard to match in various "offloading engine" - if viable DTLS offloading exists at all of course).

jukkar commented:
> Due to @jukkar's hastiness, here comes postreview, still raising some points worth attention IMHO.

I waited several days for your comments that never seemed to come, you also did not indicate that you needed more time for this. Also other people commented, Robert send new version which updated the code accordingly. After all this there is normally no need to wait more if people do not indicate they need more time.

pfalcon commented:
> I waited several days for your comments that never seemed to come, you also did not indicate that you needed more time for this. 

I didn't? https://github.com/zephyrproject-rtos/zephyr/pull/9338#issuecomment-412054498 reads "My suggestion would be not to haste with this, and let other people (not just me) to review it, e.g. until middle of next week.". Anyway, if @rlubos would want to act on some suggestions (nothing critical - beyond multi-client server support which can for sure wait), he surely can do it in other PRs.

rlubos commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -738,15 +738,14 @@ int ztls_socket(int family, int type, int proto)
-		/* Currently DTLS is not supported,
-		 * so do not allow to create datagram socket
-		 */
+	if ((proto >= IPPROTO_TLS_1_0 && proto <= IPPROTO_TLS_1_2) ||
```
I'll add extra check.

rlubos commented:
@pfalcon Thank you for the feedback, I'll apply the fixes and propose them in another PR.

> So, layering violation catches and bites us again, eh? ;-)

I wouldn't say that it's layering that bites us, but resource management. Eventlually, something like point 2 from your proposals should be implemented for DTLS. Yet, as I've found even the simple single-client DTLS implementation quite complex, I've gave up on this for now, as it would surely delay the PR and make it harder to review. Especially, that every DTLS server sample in Zephyr so far, follow this convention (`net_app`, bare mbedTLS, all use single SSL context, therefore allow to connect only one client).

But I do agree, that it's a limitation and there's a room for improvement.

pfalcon commented:
> I wouldn't say that it's layering that bites us, but resource management.

Yeah, but that's part of the layering issue - in classical model, it's flexible and all is left to application to do it in completely in different/orthogonal ways they like (at the expense of having to do it). We're trying to hide that in kernel, which saves apps from dealing with it, but now we need to find good schedule for resource management and implement it, still risking that it may be not flexible enough for some apps.

> Yet, as I've found even the simple single-client DTLS implementation quite complex, I've gave up on this for now, as it would surely delay the PR and make it harder to review.

Sure, ack on that, as I told, I barely was familiar with DTLS at all myself.

> Especially, that every DTLS server sample in Zephyr so far, follow this convention (net_app, bare mbedTLS, all use single SSL context, therefore allow to connect only one client).

I wanted to check how net_app had it, thanks for confirming that it supports just one connection either. (I remember times when net_app's TCP server supported just one connection ;-) ).

> But I do agree, that it's a limitation and there's a room for improvement.

Yeah, we just need to document it as such, and fixing that would definitely be based on the actual need to support multiple DTLS clients per server in implemented in Zephyr, which unlikely to arise soon. Actually, we may need that sooner to define DTLS offloading interface, but even that will be queued-behind TLS offloading for quite some time I guess.

Thanks for responses and the new PR, will proceed to it now!
